### ISSUE TITLE
Error while set initial estimate

### ISSUE THREAD

**Issue by JorgeFernandes-Git** [lardemua/atom#578]
Hi @miguelriemoliveira.

While trying to set the initial estimate, ATOM gave the error:

```     
  File "/home/jorge/catkin_ws/devel/lib/atom_calibration/set_initial_estimate", line 15, in <module>
    exec(compile(fh.read(), python_script, 'exec'), context)
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/scripts/set_initial_estimate", line 293, in <module>
    first_guess.init()
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/scripts/set_initial_estimate", line 111, in init
    self.sensors.append(Sensor(name, self.server, self.menu, **params))
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 88, in __init__
    self.updateAll()  # update all the transformations
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 191, in updateAll
    self.updatePreT()
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 199, in updatePreT
    self.preT = self.updateT(self.world_link, self.opt_parent_link, rospy.Time.now())
  File "/home/jorge/catkin_ws/src/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 206, in updateT
    self.listener.waitForTransform(parent_link, child_link, rospy.Time(), rospy.Duration(1.0))
  File "/opt/ros/noetic/lib/python3/dist-packages/tf/listener.py", line 76, in waitForTransform
    raise tf2_ros.TransformException(error_msg or "no such transformation: \"{}\" -> \"{}\"".format(source_frame, target_frame))
tf2.TransformException: Could not find a connection between 'zau/odom' and 'cam_1_link' because they are not part of the same tree.Tf has two or more unconnected trees.
================================================================================REQUIRED process [set_initial_estimate-8] has died!
process has died [pid 98684, exit code 1, cmd /home/jorge/catkin_ws/devel/lib/atom_calibration/set_initial_estimate -s 0.5 -f /home/jorge/catkin_ws/src/zau_bot/Zau_INESCTEC/zau_full_calibration/urdf/initial_estimate.urdf.xacro -c /home/jorge/catkin_ws/src/zau_bot/Zau_INESCTEC/zau_full_calibration/calibration/config.yml __name:=set_initial_estimate __log:=/home/jorge/.ros/log/6680be14-f64c-11ed-aca0-d727f957d12b/set_initial_estimate-8.log].
log file: /home/jorge/.ros/log/6680be14-f64c-11ed-aca0-d727f957d12b/set_initial_estimate-8*.log
Initiating shutdown!
================================================================================
```
_________________
**[Bag file](https://drive.google.com/file/d/1iz4mMn8qJKjub5x4NlGtsY3brh37jpvJ/view?usp=sharing)**

**[Zau branch](https://github.com/JorgeFernandes-Git/zau_bot/tree/Zau_INESCTEC)**

**[Config file](https://github.com/JorgeFernandes-Git/zau_bot/blob/Zau_INESCTEC/Zau_INESCTEC/zau_full_calibration/calibration/config.yml)**
_________________
Run configure:
```
rosrun atom_calibration configure_calibration_pkg -n zau_full_calibration -utf
```

Set initial estimate:
```
roslaunch zau_full_calibration set_initial_estimate.launch
```
_________________
Tf tree:
![image](https://github.com/lardemua/atom/assets/80167550/49a0c352-99b4-498f-8516-28d49d3b6fd3)

**Comment by JorgeFernandes-Git** [comment]
I tried to create a small bag file from the original, but the result was not the expected.

[Here is the small bag for test.](https://drive.google.com/file/d/17A7JByPbw-mmxIvwcVeTz6lsELEhI0sN/view?usp=sharing)

Unfortunately, this bag presents a **disconnected tf tree**:
![image](https://github.com/lardemua/atom/assets/80167550/2a046e8d-502c-4851-baac-1433949353f6)


As you know @miguelriemoliveira, we used a static_transform_publisher to publish some transformations, since the robot tf tree weren't properly configured. 
In the original bag file this seems to work, but maybe this approach can't be used for the calibration.

**Comment by manuelgitgomes** [comment]
Hello Jorge!

Have you altered the bag file in anyway?


When I did something similar, the tf_static topic seems to stop functioning.


What I did to combat it is a bit cumbersome, but I launched every static tf in a launch file.
Be sure to run it before trying to use atom.



________________________________

**Comment by JorgeFernandes-Git** [comment]
Hi @manuelgitgomes.

To divide the bag file I tried two approaches: 
* use rqt_bag, select the area I want to save and save a new bag;
* use rosbag filter, give a t.secs range and save a new bag;

Both methods resulted in an unconnected tree.

Now, I don't know if the problem is on the original bag file, or the method used to divide it.

**Comment by miguelriemoliveira** [comment]
> Unfortunately, this bag presents a disconnected tf tree:

So the problem of the disconnected tf tree is only after you get a small bag file?

Because if it also exists in the large bag file, for example in the first 30 secs, that could the reason why the set initial estimate is crashing.

In any case if you cannot create the smaller bag you have to send the link to the big one.

I will try to work on this asap.

**Comment by JorgeFernandes-Git** [comment]
> > Unfortunately, this bag presents a disconnected tf tree:
> 
> So the problem of the disconnected tf tree is only after you get a small bag file?
> 
> Because if it also exists in the large bag file, for example in the first 30 secs, that could the reason why the set initial estimate is crashing.
> 
> In any case if you cannot create the smaller bag you have to send the link to the big one.
> 
> I will try to work on this asap.

The problem is only on the smaller bag. 
In the big bag the tf tree is connected.

Here is it:
**[Bag file](https://drive.google.com/file/d/1iz4mMn8qJKjub5x4NlGtsY3brh37jpvJ/view?usp=sharing)**

**Comment by miguelriemoliveira** [comment]
Thanks. I really enjoyed the detailed summary at the beginning of the issue. Great job.

**Comment by miguelriemoliveira** [comment]
I can confirm I have the same error

```
Collecting transforms...
Traceback (most recent call last):18842   Duration: 2.021423 / 671.204832               
  File "/home/mike/workspaces/catkin_ws/devel/lib/atom_calibration/set_initial_estimate", line 15, in <module>
    exec(compile(fh.read(), python_script, 'exec'), context)
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/scripts/set_initial_estimate", line 293, in <module>
    first_guess.init()
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/scripts/set_initial_estimate", line 111, in init
    self.sensors.append(Sensor(name, self.server, self.menu, **params))
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 88, in __init__
    self.updateAll()  # update all the transformations
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 191, in updateAll
    self.updatePreT()
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 199, in updatePreT
    self.preT = self.updateT(self.world_link, self.opt_parent_link, rospy.Time.now())
  File "/home/mike/workspaces/catkin_ws/src/calibration/atom/atom_calibration/src/atom_calibration/initial_estimate/sensor.py", line 206, in updateT
    self.listener.waitForTransform(parent_link, child_link, rospy.Time(), rospy.Duration(1.0))
  File "/opt/ros/noetic/lib/python3/dist-packages/tf/listener.py", line 76, in waitForTransform
    raise tf2_ros.TransformException(error_msg or "no such transformation: \"{}\" -> \"{}\"".format(source_frame, target_frame))
tf2.TransformException: Could not find a connection between 'zau/odom' and 'cam_2_link' because they are not part of the same tree.Tf has two or more unconnected trees.
================================================================================REQUIRED process [set_initial_estimate-8] has died!
process has died [pid 22904, exit code 1, cmd /home/mike/workspaces/catkin_ws/devel/lib/atom_calibration/set_initial_estimate -s 0.5 -f /home/mike/workspaces/catkin_ws/src/calibration/robots/zau_bot/Zau_INESCTEC/zau_full_calibration/urdf/initial_estimate.urdf.xacro -c /home/mike/workspaces/catkin_ws/src/calibration/robots/zau_bot/Zau_INESCTEC/zau_full_calibration/calibration/config.yml __name:=set_initial_estimate __log:=/home/mike/.ros/log/905bf41c-f70d-11ed-b186-8dc20707cb43/set_initial_estimate-8.log].
log file: /home/mike/.ros/log/905bf41c-f70d-11ed-b186-8dc20707cb43/set_initial_estimate-8*.log
Initiating shutdown!
================================================================================

```

**Comment by JorgeFernandes-Git** [comment]
Maybe the problem is on the bag file :(

**Comment by miguelriemoliveira** [comment]
So after https://github.com/JorgeFernandes-Git/zau_bot/issues/21 I am working on this one.

The mindset is that the problem may be in the bagfile, but if ATOM brakes without an informative message it is also an ATOM problem.

When running the collect_data I get this warning:

```
[ WARN] [1684610906.945273687, 1682697331.478720122]: TF_REPEATED_DATA ignoring data with redundant timestamp for frame zau/odom at time 1682697331.408028 according to authority unknown_publisher
Warning: TF_REPEATED_DATA ignoring data with redundant timestamp for frame zau/odom at time 1682697331.408028 according to authority /rosbag_play
         at line 278 in /tmp/binarydeb/ros-noetic-tf2-0.7.6/src/buffer_core.cpp
```

So now I am going to investigate if the odom tfs are fine...

**Comment by miguelriemoliveira** [comment]
So I did a script to analyse specifically the odom tf messages. 

https://github.com/JorgeFernandes-Git/zau_bot/blob/Zau_INESCTEC/zau_bot_bringup/scripts/inspect_odom_tfs_in_bag.py

Note that when we receive a tf message it has a time stamp. Then it contains the transforms, each with is timestamp as well.

In the prints below, the Got transform refers to the delta computed with the message stamp, the second line refers to the delta computed with the transform stamp. 

```
Got transform in msg on topic /tf with parent zau/map_rn0 and child zau/odom at time 5.698397106
transform.header.stamp = 5.646736292
Got transform in msg on topic /tf with parent zau/map_rn0 and child zau/odom at time 5.737686678
transform.header.stamp = 5.700382854
Got transform in msg on topic /tf with parent zau/map_rn0 and child zau/odom at time 5.7687125
transform.header.stamp = 5.7255446469999995
Got transform in msg on topic /tf with parent zau/map_rn0 and child zau/odom at time 5.820192519
transform.header.stamp = 5.7794247389999995
```

So we can confirm that sometimes the transform stamps are repeated, which explains the warning message I was receiving. The question is whether or not this is causing the error in the set_initial_estimate.

Also, I am not sure why this occurs ...

**Comment by miguelriemoliveira** [comment]
Hi @JorgeFernandes-Git ,

still exploring this.  I found out that the problem is only with sensor cam_2. Sensor cam_1 works fine on set_initial_estimate.
I looked into the tf tree again and found this:

![image](https://github.com/lardemua/atom/assets/11350998/4c540faf-f344-498a-97b4-f1674672549c)

It is strange but the connection from zau/base_link to cam_2_link is the only that has a 10Hz average rate.

This is not a static transform, as those are marked with a 10000 average rate (see zau/base_link to zau/base_laser_link_0 for example).

Do you know how you published this specific transform?

**Comment by miguelriemoliveira** [comment]
Did another experiment. Prepared a script that would produce a new bagfile with messages only when two specific tfs are received:

1. zau/map_rn0 to zau/odom
2. zau/base_link to cam_2_link

This is just so we can know when these tf transformations are received. Here are the results:

![image](https://github.com/lardemua/atom/assets/11350998/6cb72090-1337-4cf0-8f7d-af79d4f63e15)

It looks good, there are no gaps, these transforms arrive frequently without interruptions.

Still searching ...

**Comment by miguelriemoliveira** [comment]
Found where the problem is. It has to do with the additional_tfs. When these lines are commented, the code works fine.

https://github.com/lardemua/atom/blob/d3828f7b9e1fcbe008b760556b77609b18833a08/atom_calibration/scripts/set_initial_estimate#L115-L131

**Comment by miguelriemoliveira** [comment]
Hi @JorgeFernandes-Git ,

finally I found the problem. It was in the additional_tfs code. We tested putting some sleep time but in the sensor code, not the additional tfs. I fixed by creating a shared transform listener to all.

There is another problem now, which occurs when the additional tf transform interactive marker is moved.
The problem is that the arm starts shaking around. We see it oscillating between some of these images below.

![image](https://github.com/lardemua/atom/assets/11350998/09dd67df-a61c-482d-9f9a-f1070e14672e)

![image](https://github.com/lardemua/atom/assets/11350998/a9ad40d3-6571-4e7d-9232-56eef2d3a838)

![image](https://github.com/lardemua/atom/assets/11350998/aa689970-2f9f-49b3-b708-b44cf7682b10)

The problem here I think is that the  tf that corresponds to the additional_tf (zau/base_link to base_link) should be static and its not. Because of this, when we move the interactive marker, there are two entities continuously publishing this transformation:

1. the bag file contains there transforms 
2. The set_initial_estimate code

**Comment by miguelriemoliveira** [comment]
I think this is the end of the line. I mean, you can try to repair the bag, by removing these transforms and adding a static one for example. @JorgeFernandes-Git if you do that let me know how it went.

Will make a check during the package configuration. Opened #581 for this.

**Comment by miguelriemoliveira** [comment]
Hi @JorgeFernandes-Git ,

I added a verification in the configuration that checks if the transformations to be optimized are static. 
The configuration fails they are not static.

Now, the zaubot bagfile fails the configuration because the zau/base_link to base_link transformation is not static and it is listed in the additional_tfs as a transformation to optimize.

The configuration now says:

`Error: additional_tf  zau/base_link_to_base_link defines transformation zau/base_link to base_link to be calibrated but this transformation is not static and therefore cannot be calibrated.`

I am closing this issue as the problem has been identified. If you want to calibrate using that bagfile you should somehow convert it to change the transform zau/base_link to base_link to static.

If you need help for that please create a separate issue.

**Comment by miguelriemoliveira** [comment]
> I am closing this issue as the problem has been identified. If you want to calibrate using that bagfile you should somehow convert it to change the transform zau/base_link to base_link to static.

We could do a generic script for this to put in ATOM, what do you think?

**Comment by miguelriemoliveira** [comment]
Solution in https://github.com/JorgeFernandes-Git/zau_bot/issues/21
